### ISSUE TITLE
Parser parses #IMPLIED use attributes as NORMAL (and vice versa)

### ISSUE THREAD

**Issue by raner** [eclipse-ee4j/jaxb-dtd-parser#39]
I believe that there is a fundamental flaw in the `DTDParser`'s attribute use parsing. Using various test DTDs, it appears that attributes with `#IMPLIED` default declarations are assigned attribute use `USE_NORMAL`. Vice versa, attributes with a declared default value (and neither `#IMPLIED`, `#REQUIRED` nor `#FIXED` designation) do appear as `USE_IMPLIED`.

The following test case, I believe, demonstrates this issue with a very simple DTD:
```
import java.io.StringReader;
import java.util.HashMap;
import java.util.Map;
import com.sun.xml.dtdparser.DTDParser;
import com.sun.xml.dtdparser.DTDHandlerBase;
import com.sun.xml.dtdparser.DTDEventListener;
import org.junit.Test;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import static org.junit.Assert.assertEquals;

public class AttributeUseTest
{
  Map<String, Short> attributes = new HashMap<>();

  public AttributeUseTest() throws Exception
  {
    String dtd =
    """
      <!ELEMENT root EMPTY>
      <!ATTLIST root
        normal CDATA "default"
        implied CDATA #IMPLIED
        required CDATA #REQUIRED
        fixed CDATA #FIXED "default"
      >
    """;
    DTDEventListener handler = new DTDHandlerBase()
    {
      @Override
      public void attributeDecl(String element, String name, String type, String[] enums, short use, String defaultValue) throws SAXException
      {
          attributes.put(name, use);
      }
    };
    InputSource input = new InputSource(new StringReader(dtd));
    DTDParser parser = new DTDParser();
    parser.setDtdHandler(handler);
    parser.parse(input);
  }

  @Test
  public void testNormalAttribute()
  {
    assertEquals(DTDEventListener.USE_NORMAL, attributes.get("normal").shortValue());
  }

  @Test
  public void testImpliedAttribute()
  {
    assertEquals(DTDEventListener.USE_IMPLIED, attributes.get("implied").shortValue());
  }

  @Test
  public void testRequiredAttribute()
  {
      assertEquals(DTDEventListener.USE_REQUIRED, attributes.get("required").shortValue());
  }

  @Test
  public void testFixedAttribute()
  {
      assertEquals(DTDEventListener.USE_FIXED, attributes.get("fixed").shortValue());
  }
}
```
The test cases for `#REQUIRED` and `#FIXED` pass as expected, but the two other test cases both fail.

As I ran this in the debugger, I noticed some strange looking code in `DTDParser.maybeAttlistDecl()`:

https://github.com/eclipse-ee4j/jaxb-dtd-parser/blob/c9b1bb6b20eee9452561373faba867265353e1cc/dtd-parser/src/main/java/com/sun/xml/dtdparser/DTDParser.java#L1517-L1518

Intuitively, the  negation `!peek("#IMPLIED")` (instead of just `peek("#IMPLIED")`) seems incorrect here, since it would select `USE_IMPLIED` specifically if `#IMPLIED` was **not** present. This explains why attributes without any default declarations end up being recognized as implied.

Further down, `USE_NORMAL` is assigned as a fallback value:

https://github.com/eclipse-ee4j/jaxb-dtd-parser/blob/c9b1bb6b20eee9452561373faba867265353e1cc/dtd-parser/src/main/java/com/sun/xml/dtdparser/DTDParser.java#L1538-L1540

This explains why `#IMPLIED` attributes (which do not get recognized further up, due to the negation) are designated as `USE_NORMAL`.

**Comment by raner** [comment]
It would be great if this could be fixed in both the 1.4.x and 1.5.x branches. Personally, I need this for a project that is still on Java 8, and since the 1.5.0 release requires Java 11, I am using version 1.4.5. The described problem can be reproduced both in 1.4.5 and 1.5.0.

**Comment by antoniosanct** [comment]
@raner 
You are right about this. Based on [this](https://www.w3.org/TR/WD-html40-970708/intro/sgmltut.html),

_Whether the default value of the attribute is implicit (keyword "#IMPLIED"), in which case **the default value must be supplied by the user agent** (in some cases via inheritance from parent elements); always required (keyword "#REQUIRED"); or fixed to the given value (keyword "#FIXED"). Some attributes explicitly specify a default value for the attribute._ 

Then, the right decision would be:
- delete the bug boolean comparation deleting '!' in 1571 and
- move block 1521-1533 after USE_NORMAL declaration

@lukasj Is it useful to create a PR for fix this issue?

Regards,
Antonio.
